src/dpdk_statefulset_updater.py

```diff
@@ -0,0 +1,175 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module used to update Kubernetes Statefulset to support DPDK."""
+
+import logging
+from typing import Iterable, Optional
+
+from lightkube import Client
+from lightkube.core.exceptions import ApiError
+from lightkube.models.core_v1 import Container
+from lightkube.resources.apps_v1 import StatefulSet
+
+logger = logging.getLogger(__name__)
+
+
+class DPDKStatefulSetUpdaterError(Exception):
+    """DPDKStatefulSetUpdaterError."""
+
+    def __init__(self, message: str):
+        self.message = message
+        super().__init__(self.message)
+
+
+class DPDKStatefulSetUpdater:
+    """Class used to update Kubernetes Statefulset to support DPDK."""
+
+    def __init__(
+        self,
+        statefulset_name: str,
+        namespace: str,
+        dpdk_access_interface_resource_name: str,
+        dpdk_core_interface_resource_name: str,
+    ):
+        self.k8s_client = Client()
+        self.statefulset_name = statefulset_name
+        self.namespace = namespace
+        self.dpdk_resource_requirements = {
+            "requests": {
+                dpdk_access_interface_resource_name: "1",
+                dpdk_core_interface_resource_name: "1",
+            },
+            "limits": {
+                dpdk_access_interface_resource_name: "1",
+                dpdk_core_interface_resource_name: "1",
+            },
+        }
+
+    def container_configured_for_dpdk(self, container_name: str) -> bool:
+        """Checks whether the container config required for DPDK has been applied or not.
+
+        Args:
+            container_name (str): Name of the container to update
+
+        Returns:
+            bool: True if container config required for DPDK is applied, otherwise False
+        """
+        statefulset = self._get_statefulset(self.statefulset_name, self.namespace)
+        if not statefulset:
+            raise RuntimeError("StatefulSet not found!")
+        container = self._get_container(
+            container_name=container_name,
+            containers=statefulset.spec.template.spec.containers,
+        )
+        if not container:
+            raise RuntimeError("Container not found!")
+        if not container.securityContext.privileged:
+            return False
+        if not self._resource_requirements_applied(container, self.dpdk_resource_requirements):
+            return False
+        return True
+
+    def configure_container_for_dpdk(self, container_name: str) -> None:
+        """Applies config required by DPDK to a given container.
+
+        Args:
+            container_name (str): Name of the container to update
+        """
+        statefulset = self._get_statefulset(self.statefulset_name, self.namespace)
+        if not statefulset:
+            raise RuntimeError("StatefulSet not found!")
```

Comment:
Why do we raise RuntimeError instead of DPDKStatefulSetUpdaterError

Reply:
Because this should never happen. `_get_statefulset` will raise `DPDKStatefulSetUpdaterError` when appropriate. This check is in fact just to make static analysis happy.
Alternatively, I could do 
```python
assert statefulset
```